Introduce mixed_sign_less()/min()/max()

alexkaratarakis opened the pull request:
Previously, there was mixed sign comparison:
```
reflected_min_v = -120 > 0U ? -120 : 0
```
which returns -120 because the condition is true, since the
compiler casts the signed value to unsigned, leading to a huge value.
This caused the following values as result:

```
reflected_min_v == -120
min_v = -1
static_cast<U>(min_v) = 18446744073709551615
```

The last one is used in magic_enum::enum_index().

The huge value was caught by static analysis (coverity), so this might get flagged by others.
This issue was introduced [here](https://github.com/Neargye/magic_enum/commit/370a1d523684532eb743ec78213fd323f4b011f3)

alexkaratarakis committed:
Introduce mixed_sign_less()/min()/max()

Previously, there was mixed sign comparison:
```
reflected_min_v = -120 > 0U ? -120 : 0
```
which returns -120 because the condition is true, since the
compiler casts the signed value to unsigned, leading to a huge value.
This caused the following values as result:

```
reflected_min_v == -120
min_v = -1
static_cast<U>(min_v) = 18446744073709551615
```

The last one is used in magic_enum::enum_index().

Neargye commented on include/magic_enum.hpp:
```diff
@@ -141,6 +141,34 @@ constexpr std::string_view pretty_name(std::string_view name) noexcept {
+  }
+  else if constexpr(std::is_signed_v<L>) {
+    // if 'left' is negative, then result is 'true', otherwise cast & compare
+    return left < 0 || static_cast<R>(left) < right;
```
How about `return left < 0 || static_cast<std::make_unsigned_t<L>>(left) < right;`
For fix it L::max() > R::max()

Neargye commented on include/magic_enum.hpp:
```diff
@@ -141,6 +141,34 @@ constexpr std::string_view pretty_name(std::string_view name) noexcept {
+  }
+  else { // std::is_signed_v<R>
+    // if 'right' is negative, then result is 'false', otherwise cast & compare
+    return right >= 0 && left < static_cast<L>(right);
```
Same `return right >= 0 && left < static_cast<std::make_unsigned_t<R>>(right);`

Neargye commented:
This is a really good catch.

Neargye commented:
How about return left < 0 || static_cast<std::make_unsigned_t<L>>(left) < right;
For fix it L::max() > R::max()

alexkaratarakis commented:
Good point, there might be narrowing with the current `static_cast<R>`. I will test with `std::make_unsigned`.

alexkaratarakis commented:
Indeed there was a bug with `L::max() > R::max()`.
Fixed and added tests for this.

These were failing before:
```C++
REQUIRE(!mixed_sign_less(std::numeric_limits<int64_t>::max() - 17, std::numeric_limits<uint32_t>::max()));
REQUIRE(mixed_sign_less(std::numeric_limits<uint32_t>::max(), std::numeric_limits<int64_t>::max()));
REQUIRE(mixed_sign_less(std::numeric_limits<uint32_t>::max(), std::numeric_limits<int64_t>::max() - 17));
```

Interestingly, this was succeeding:
```C++
REQUIRE(!mixed_sign_less(std::numeric_limits<int64_t>::max(), std::numeric_limits<uint32_t>::max()));
```
because:
```C++
static_cast<uint32_t>(std::numeric_limits<int64_t>::max()) == std::numeric_limits<uint32_t>::max()
```
This corner case is why I added the '17' offset.
